File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        rows = len(matrix)
        for i in range(rows//2):
            for j in range(i, rows-i-1):
                temp0 = matrix[i][j]
                temp1 = matrix[j][rows-i-1]
                temp2 = matrix[rows-i-1][rows-j-1]
                temp3 = matrix[rows-j-1][i]

                matrix[j][rows-i-1] = temp0
                matrix[rows-i-1][rows-j-1] = temp1
                matrix[rows-j-1][i] = temp2
                matrix[i][j] = temp3

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        transpose(matrix);
        reflect(matrix); 
    }
private:
    void transpose(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
    }
    void reflect(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i < n; i++) {
            for (int j=0; j < n / 2; j++) {
                swap(matrix[i][j], matrix[i][n-1-j]);
            }
        }
    }
};